h2/src/main/org/h2/engine/Session.java

```diff
@@ -542,11 +554,8 @@ public Command prepareLocal(String sql) {
                     "session closed");
         }
         Command command;
-        if (queryCacheSize > 0) {
-            if (queryCache == null) {
-                queryCache = SmallLRUCache.newInstance(queryCacheSize);
-                modificationMetaID = database.getModificationMetaId();
-            } else {
+        if (queryCache != null) {
+            synchronized (queryCache) {
```

Comment:
Yes, I double checked, you are right. This method should not be executed concurrently. I'll revert this part.